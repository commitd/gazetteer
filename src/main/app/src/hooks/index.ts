export * from './useDialog'
export * from './useToggle'
