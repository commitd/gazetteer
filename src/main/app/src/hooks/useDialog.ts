import { useState } from 'react'

/**
 * Utility hook for open and close of dialogs
 *
 * <p> returns the state value, a open function, and a close function.
 */
export function useDialog(): [boolean, () => void, () => void] {
  const [value, setValue] = useState(false)

  const open = (): void => setValue(true)
  const close = (): void => setValue(false)

  return [value, open, close]
}
