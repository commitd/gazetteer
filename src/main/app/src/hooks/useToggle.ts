import { useState } from 'react'

/**
 * Utility hook for boolean toggle operations
 *
 * <p> returns the value, a toggle function, and the original setValue in case required.
 */
export function useToggle(
  startState: boolean
): [boolean, () => void, (force: boolean) => void] {
  const [value, setValue] = useState(startState)

  const toggleValue = (): void => setValue(!value)

  return [value, toggleValue, setValue]
}
