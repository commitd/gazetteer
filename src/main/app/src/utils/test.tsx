import { StylesProvider } from '@material-ui/styles'
import {
  createHistory,
  createMemorySource,
  History,
  LocationProvider,
} from '@reach/router'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { GenerateId } from 'jss'
import React from 'react'
import { ThemeProvider } from '@committed/components'

const generateClassName: GenerateId = (rule, styleSheet) => {
  const prefix =
    styleSheet === undefined || styleSheet.options.classNamePrefix === undefined
      ? ''
      : styleSheet.options.classNamePrefix
  return `${prefix}-${rule.key}`
}

const LightTheme: React.FC<{ readonly children?: React.ReactNode }> = ({
  children,
}) => (
  <StylesProvider generateClassName={generateClassName}>
    <ThemeProvider choice="light">{children}</ThemeProvider>
  </StylesProvider>
)

const DarkTheme: React.FC<{ readonly children?: React.ReactNode }> = ({
  children,
}) => (
  <StylesProvider generateClassName={generateClassName}>
    <ThemeProvider choice="dark">{children}</ThemeProvider>
  </StylesProvider>
)

export const renderPlain = render

export const renderLight = (
  ui: Readonly<React.ReactElement>,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => render(ui, { wrapper: LightTheme, ...options })

export const renderDark = (
  ui: Readonly<React.ReactElement>,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => render(ui, { wrapper: DarkTheme, ...options })

export function renderWithRouter(
  ui: Readonly<React.ReactElement>,
  route = '/',
  options?: Omit<RenderOptions, 'queries'>,
  delegate = renderLight
): RenderResult & { history: History } {
  const history = createHistory(createMemorySource(route))
  return {
    ...delegate(
      <LocationProvider history={history}>{ui}</LocationProvider>,
      options
    ),
    history,
  }
}
// re-export everything
export * from '@testing-library/react'
// override render method
export { renderLight as render }
export { userEvent }
