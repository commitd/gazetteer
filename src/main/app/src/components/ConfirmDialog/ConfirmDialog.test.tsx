import React from 'react'
import { Button } from '@committed/components'
import { render, renderDark, userEvent } from '../../utils/test'
import { ConfirmDialog } from '.'
import { useToggle } from '../../hooks'

it('renders without error', () => {
  const { asFragment } = render(
    <ConfirmDialog
      open={true}
      onConfirm={(): void => undefined}
      onClose={(): void => undefined}
    />
  )
  expect(asFragment()).toMatchSnapshot()
})

it('renders dark without error', () => {
  const { asFragment } = renderDark(
    <ConfirmDialog
      open={true}
      onConfirm={(): void => undefined}
      onClose={(): void => undefined}
    />
  )
  expect(asFragment()).toMatchSnapshot()
})

const CanConfirm: React.FC<{ title: string; onConfirm: () => void }> = ({
  title,
  onConfirm,
}) => {
  const [open, toggleOpen] = useToggle(false)

  return (
    <>
      <Button onClick={toggleOpen}>Open</Button>
      <ConfirmDialog
        open={open}
        title={title}
        onClose={toggleOpen}
        onConfirm={onConfirm}
      />
    </>
  )
}

it('can be cancelled', () => {
  const title = 'Do you confirm?'
  const onConfirm = jest.fn()
  const { getByText, getByRole } = render(
    <CanConfirm title={title} onConfirm={onConfirm} />
  )

  userEvent.click(getByRole('button', { name: /Open/i }))
  getByText(title)
  userEvent.click(getByRole('button', { name: /Cancel/i }))
  expect(onConfirm).not.toHaveBeenCalled()
})

it('can confirm', () => {
  const title = 'Do you confirm?'
  const onConfirm = jest.fn()
  const { getByText, getByRole } = render(
    <CanConfirm title={title} onConfirm={onConfirm} />
  )

  userEvent.click(getByRole('button', { name: /Open/i }))
  getByText(title)
  userEvent.click(getByRole('button', { name: /Confirm/i }))
  expect(onConfirm).toHaveBeenCalled()
})
