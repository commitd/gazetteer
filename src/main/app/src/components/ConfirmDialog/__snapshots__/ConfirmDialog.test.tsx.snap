// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders dark without error 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`renders without error 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;
